erigon-lib/commitment/commitment.go

```diff
@@ -758,3 +758,230 @@ func DecodeBranchAndCollectStat(key, branch []byte, tv TrieVariant) *BranchStat
 	}
 	return stat
 }
+
+// Defines how to evaluate commitments
+type Mode uint
+
+const (
+	ModeDisabled Mode = 0
+	ModeDirect   Mode = 1
+	ModeUpdate   Mode = 2
+)
+
+func (m Mode) String() string {
+	switch m {
+	case ModeDisabled:
+		return "disabled"
+	case ModeDirect:
+		return "direct"
+	case ModeUpdate:
+		return "update"
+	default:
+		return "unknown"
+	}
+}
+
+func ParseCommitmentMode(s string) Mode {
+	var mode Mode
+	switch s {
+	case "off":
+		mode = ModeDisabled
+	case "update":
+		mode = ModeUpdate
+	default:
+		mode = ModeDirect
+	}
+	return mode
+}
+
+type UpdateTree struct {
+	tree   *btree.BTreeG[*KeyUpdate]
+	keccak cryptozerocopy.KeccakState
+	keys   map[string]struct{}
+	mode   Mode
+}
+
+func NewUpdateTree(m Mode) *UpdateTree {
+	return &UpdateTree{
+		tree:   btree.NewG[*KeyUpdate](64, keyUpdateLessFn),
+		keccak: sha3.NewLegacyKeccak256().(cryptozerocopy.KeccakState),
+		keys:   map[string]struct{}{},
+		mode:   m,
+	}
+}
+
+func (t *UpdateTree) get(key []byte) (*KeyUpdate, bool) {
+	c := &KeyUpdate{plainKey: key, update: Update{CodeHashOrStorage: EmptyCodeHashArray}}
+	el, ok := t.tree.Get(c)
+	if ok {
+		return el, true
+	}
+	c.plainKey = common.Copy(c.plainKey)
+	return c, false
+}
+
+// TouchPlainKey marks plainKey as updated and applies different fn for different key types
+// (different behaviour for Code, Account and Storage key modifications).
+func (t *UpdateTree) TouchPlainKey(key string, val []byte, fn func(c *KeyUpdate, val []byte)) {
+	switch t.mode {
+	case ModeUpdate:
+		item, _ := t.get([]byte(key))
+		fn(item, val)
+		t.tree.ReplaceOrInsert(item)
+	case ModeDirect:
+		t.keys[key] = struct{}{}
+	default:
+	}
+}
+
+func (t *UpdateTree) Size() uint64 {
+	return uint64(len(t.keys))
+}
+
+func (t *UpdateTree) TouchAccount(c *KeyUpdate, val []byte) {
+	if len(val) == 0 {
+		c.update.Flags = DeleteUpdate
+		return
+	}
+	if c.update.Flags&DeleteUpdate != 0 {
+		c.update.Flags ^= DeleteUpdate
+	}
+	nonce, balance, chash := types.DecodeAccountBytesV3(val)
+	if c.update.Nonce != nonce {
+		c.update.Nonce = nonce
+		c.update.Flags |= NonceUpdate
+	}
+	if !c.update.Balance.Eq(balance) {
+		c.update.Balance.Set(balance)
+		c.update.Flags |= BalanceUpdate
+	}
+	if !bytes.Equal(chash, c.update.CodeHashOrStorage[:]) {
+		if len(chash) == 0 {
+			c.update.ValLength = length.Hash
+			copy(c.update.CodeHashOrStorage[:], EmptyCodeHash)
+		} else {
+			copy(c.update.CodeHashOrStorage[:], chash)
+			c.update.ValLength = length.Hash
+			c.update.Flags |= CodeUpdate
+		}
+	}
+}
+
+func (t *UpdateTree) UpdatePrefix(prefix, val []byte, fn func(c *KeyUpdate, val []byte)) {
+	t.tree.AscendGreaterOrEqual(&KeyUpdate{}, func(item *KeyUpdate) bool {
+		if !bytes.HasPrefix(item.plainKey, prefix) {
+			return false
```

Comment:
let's remove this func. it's not used and it seems broken - because it stops iteration even if such prefix exists (`KeyUpdate{}` does full-scan)